### PR TITLE
GitStatusMonitor: Avoid background updates if GUI is not visible

### PR DESCRIPTION
**Pull request by gerhardol**
Part of #6521

## Proposed changes

Avoid running git-status when GE Browse is minimized

## Test methodology
Check Git Command Log

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
Should we also disable it if the app is not active? And resume when the app becomes active?

File: GitCommands/Utils/EnvUtils.cs [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
 using System.Diagnostics;
+using System.Runtime.InteropServices;
 using Microsoft.Win32;
 
 namespace GitCommands.Utils
 {
     public static class EnvUtils
     {
+        [DllImport("user32.dll")]
+        private static extern bool IsIconic(IntPtr handle);
```
Comment (RussKie) [inline GitCommands/Utils/EnvUtils.cs:11]: We define interops in the UI layer, have a look at GitUI/Interop.

**Comment by gerhardol** [comment]
> Should we also disable it if the app is not active? And resume when the app becomes active?

No. FileSystemWatcher is paused when there is a change and the time still fires every 100 ms, so basically only the IsIconic() call would be not be needed.
(I would like to change the start/stop of GitStatusMonitor in a similar way to avoid the initial git-status when not really needed, but that is a later change.)

**Comment by mstv** [comment]
> Should we also disable it if the app is not active? And resume when the app becomes active?

If you use the FileSystemWatcher, then GE is expected to update if it is visible.

**Comment by gerhardol** [comment]
> If you use the FileSystemWatcher, then GE is expected to update if it is visible.

This change suspends the git-status call, when no longer minimized, git-status resumes (and FileSystemWatcher events are enabled again)

File: GitCommands/Utils/EnvUtils.cs (hunk 1 above)
Comment (gerhardol) [reply GitCommands/Utils/EnvUtils.cs]: I added Interops, should probably be added to GitCommands instead (Interops only in GitUI now)

**Comment by RussKie** [comment]
If we suppose that GitUI is Windows only and GitCommands could be portable,
then interops belong to the UI layer.

**Comment by gerhardol** [comment]
> If we suppose that GitUI is Windows only and GitCommands could be portable,
then interops belong to the UI layer.

Add a UIEnv then? Or keep in GitStatusMonitor for now (TODO)?

File: GitCommands/Utils/EnvUtils.cs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.Diagnostics;
+using System.Runtime.InteropServices;
```
Comment (RussKie) [inline GitCommands/Utils/EnvUtils.cs:3]: We can revert this change

File: GitUI/CommandsDialogs/BrowseDialog/GitStatusMonitor.cs [hunk 3]
```diff
@@ -309,16 +311,29 @@ private void StartWatchingChanges(string workTreePath, string gitDirPath)
...
-            if (CurrentStatus != GitStatusMonitorState.Running)
+            var currentProcess = Process.GetCurrentProcess();
+            if (currentProcess is null)
             {
-                return;
+                return false;
             }
 
-            if (_nextUpdateTime - Environment.TickCount > 0)
+            return NativeMethods.IsIconic(currentProcess.MainWindowHandle) == Interop.BOOL.TRUE;
```
Comment (RussKie) [inline GitUI/CommandsDialogs/BrowseDialog/GitStatusMonitor.cs:327]: We also have `.IsTrue()` extension
```suggestion
            return NativeMethods.IsIconic(currentProcess.MainWindowHandle).IsTrue();
```

**Comment by RussKie** [comment]
> Add a UIEnv then? Or keep in GitStatusMonitor for now (TODO)?

I think what you have currently is good.
Unless you're asking about something else...

**Comment by gerhardol** [comment]
> I think what you have currently is good.
Unless you're asking about something else...

Keeping then (the question was if to separate this environment specific to a separate module).
Rebased too.

**Comment by gerhardol** [comment]
@msftbot merge in 24 hours
